auction/contracts/AuctionHouse.sol

```diff
@@ -0,0 +1,333 @@
+                payout[0].account = payable(seller);
+                payout[0].value = 10000;
+                aucData.payouts = payout;
+            }
+            doTransfersWithFees(
+                currentAuction.lastBid.amount, 
+                address(this), 
+                MatchFees(currentAuction.protocolFee, currentAuction.protocolFee, LibFeeSide.FeeSide.TAKE),
+                bidData, 
+                aucData, 
+                currentAuction.buyAsset, 
+                currentAuction.sellAsset.assetType, 
+                TO_MAKER
+            );
+            transferPayouts(currentAuction.sellAsset.assetType, currentAuction.sellAsset.value, address(this), bidData.payouts, TO_TAKER);
+        } else {
+            transfer(currentAuction.sellAsset, address(this), seller, TO_SELLER, UNLOCK); //nft back to seller
+        }
+    }
+
+    /// @dev returns true if auction started and hasn't finished yet, false otherwise
+    function checkAuctionRangeTime(uint _auctionId) public view returns (bool){
+        uint currentTime = block.timestamp;
+        LibAucDataV1.DataV1 memory aucData = LibAucDataV1.parse(auctions[_auctionId].data, auctions[_auctionId].dataType);
+        if (aucData.startTime > 0 && aucData.startTime > currentTime) {
+            return false;
+        }
+        uint endTime = auctions[_auctionId].endTime;
+        if (endTime > 0 && endTime < currentTime){
+            return false;
+        }
+        
+        return true;
+    }
+
+    /// @dev deletes auction after finalizing
+    function deactivateAuction(uint _auctionId) internal {
+        emit AuctionFinished(_auctionId, auctions[_auctionId]);
+        deleteAuctionForToken(auctions[_auctionId].sellAsset);
+        delete auctions[_auctionId];
+    }
+
+    /// @dev cancels existing auction without bid
+    function cancel(uint _auctionId) external {
+        require(checkAuctionExistence(_auctionId), "there is no auction with this id");
+        Auction storage currentAuction = auctions[_auctionId];
+        address seller = currentAuction.seller;
+        require(seller == _msgSender(), "auction owner not detected");
+        address buyer = currentAuction.buyer;
+        require(buyer == address(0), "can't cancel auction with bid");
+        transfer(currentAuction.sellAsset, address(this), seller, TO_SELLER, UNLOCK); //nft transfer back to seller
+        deactivateAuction(_auctionId);
+        AuctionCancelled(_auctionId);
+    }
+
+    /// @dev buyout auction if bid satisfies buyout condition
+    function buyOut(uint _auctionId, Bid memory bid) external payable {
+        checkAuctionInProgress(_auctionId);
+
+        require(buyOutVerify(_auctionId, bid.amount), "not enough for buyout auction");
+        _buyOut(_auctionId, bid);
+        doTransfers(_auctionId);
+        deactivateAuction(_auctionId);
+    }
+
+    /// @dev makes buyout bid the current bid
+    function _buyOut(uint _auctionId, Bid memory bid) internal {
```

Comment:
when we call _buyOut we always call "doTransfers" next.
so we always transfer buyAssets from bidder to contract and transfer it to seller after. In this case transfer to contract is unnecessary and costly. Could we cut of it, by chance? Could we create a special method like doTransfer, which will transfer buyAsset directly from bidder to seller (and sellAsset from contract to bidder - it's the same as in doTransfers)? method could be called _buyOut - buy out auction with buyAsset directly

Reply:
that is a great idea, i tried to fix it but only broke everything. created a task https://rarible.atlassian.net/jira/software/projects/RPC/boards/83/backlog?selectedIssue=RPC-141 and will take a look when we will be refactoring auction after merging splitting exchangev2 and rarible transfer manager into 2 separate contracts